Add warning for large samples in PEC

sid1993 opened the pull request:
Fixes #454

[More details of the fix.](https://github.com/unitaryfund/mitiq/issues/454#issuecomment-735456307)

- Added test for large sample in `test_pec.py`

- Tested UserWarning by removing its filter in `pytest.ini`.
![warning](https://user-images.githubusercontent.com/4842078/100599911-9ebea480-3326-11eb-9c35-d9099197e179.png)

andreamari commented on mitiq/pec/tests/test_pec.py:
```diff
@@ -110,7 +110,7 @@ def test_execute_with_pec_with_different_samples(circuit: Circuit, seed):
 
 
-@mark.parametrize("num_samples", [100, 1000])
+@mark.parametrize("num_samples", [100, 1000, 100001])
```
Instead of changing an existing test, I would add a new test dedicated to the new warning.
There is a specific way to test that warnings, for example you can have a look at this similar test: https://github.com/unitaryfund/mitiq/blob/e96e8a1ce50a9a70f215cd35c3ce24957a60e0d7/mitiq/zne/tests/test_inference.py#L605

andreamari commented on mitiq/pec/pec.py:
```diff
@@ -102,6 +103,13 @@ def execute_with_pec(
+    # Issue warning for very large sample size
+    if num_samples > 10 ** 5:
+        warn(
+            "Sample size in PEC is very large.It may take serveral mintues.",
```
```suggestion
            "The number of PEC samples is very large. It may take several minutes."
            " It may be necessary to reduce 'precision' or 'num_samples'.",
```
We could add this suggestion for the user. 

I apologize if this change creates flake8 problems. If so, they should be easy to fix.

andreamari reviewed (commented):
Thanks @sid1993 !
I left a couple of suggestions.
Feel free to also add a new line on the CHANGELOG.md file about this PR.

sid1993 commented:
- Updated CHANGELOG.md
- Successfully tested warning for large samples in PEC
![Success_Warning](https://user-images.githubusercontent.com/4842078/100663956-01db2600-337d-11eb-9321-dcfddce6c78b.png)
- Failure tests for no warning and incorrect warning category/message :
![Failure_Test_Wrong_Category](https://user-images.githubusercontent.com/4842078/100664210-52528380-337d-11eb-8dc2-75c3b27e3ac6.png)
![Failure_Incorrect_Message](https://user-images.githubusercontent.com/4842078/100664225-55e60a80-337d-11eb-8d75-00606275a1d3.png)

- Added only one test case in `test_large_sample_size_warning` as each of them is increasing the overall testing time by 1.5-2 mins. If more test cases are required, I would suggest to decrease the warning threshold to `10 ** 4`.

andreamari commented:
@sid1993 Thanks!
I am a bit concerned about using a 2 min test just for the warning. Is there a way to test the warning as you did but abort the execution after the warning is raised? Maybe with some timeout function or any other method?

sid1993 commented:
> @sid1993 Thanks!
> I am a bit concerned about using a 2 min test just for the warning. Is there a way to test the warning as you did but abort the execution after the warning is raised? Maybe with some timeout function or any other method?

@andreamari : Yes, I am trying to figure out a way and will update you soon.

sid1993 commented:
> > @sid1993 Thanks!
> > I am a bit concerned about using a 2 min test just for the warning. Is there a way to test the warning as you did but abort the execution after the warning is raised? Maybe with some timeout function or any other method?
> 
> @andreamari : Yes, I am trying to figure out a way and will update you soon.

@andreamari : Funny story :) I had multiple VMs up on my system when i tested it and that led to incorrect results. Test time is fine just  5-6 secs more.
![Overall_Test_Time](https://user-images.githubusercontent.com/4842078/100922645-24e41200-3504-11eb-9803-8c6aa25ffda5.png)

sid1993 commented:
@andreamari : I have added another test case also. Everything looks good.

andreamari reviewed (approved):
Great! Thanks @sid1993 for your nice contribution!

crazy4pi314 commented:
@all-contributors please add @sid1993 for code, bug
